SlevomatCodingStandard/Sniffs/Namespaces/ReferenceUsedNamesOnlySniff.php

```diff
@@ -125,10 +129,33 @@ public function process(\PHP_CodeSniffer_File $phpcsFile, $openTagPointer)
 		$tokens = $phpcsFile->getTokens();
 
 		$referencedNames = ReferencedNameHelper::getAllReferencedNames($phpcsFile, $openTagPointer);
+
+		if ($this->allowFQNForCollidingClasses) {
+			$referencesIndex = array_combine(
+				array_map(
+					function (ReferencedName $referencedName) {
+						return $referencedName->getNameAsReferencedInFile();
+					},
+					$referencedNames
+				),
+				$referencedNames
+			);
+			$classPointer = $phpcsFile->findNext(TokenHelper::$typeKeywordTokenCodes, 0);
+			$thisClassName = $tokens[TokenHelper::findNextEffective($phpcsFile, $classPointer + 1)]['content'];
+		}
+
 		foreach ($referencedNames as $referencedName) {
 			$name = $referencedName->getNameAsReferencedInFile();
 			$nameStartPointer = $referencedName->getStartPointer();
 			$canonicalName = NamespaceHelper::normalizeToCanonicalName($name);
+
+			if ($this->allowFQNForCollidingClasses) {
+				$unqualifiedClassName = NamespaceHelper::getUnqualifiedNameFromFullyQualifiedName($referencedName->getNameAsReferencedInFile());
+				if (isset($referencesIndex[$unqualifiedClassName]) || (isset($thisClassName) ? $unqualifiedClassName === $thisClassName : false)) {
```

Comment:
`isset($thisClassName) && $thisClassName === $unqualifiedClassName` looks better to me.